File: poclb/logit.c
/*
 * Copyright (c) 2020 NVI, Inc.
 *
 * This file is part of VLBI Field System
 * (see http://github.com/nvi-inc/fs).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
/* logit

   logit formats message and errors for logging and sends the
   buffer to ddout.
*/

#include <string.h>
#include <stdio.h>

#include "../include/params.h"
#include "../include/fs_types.h"
#include "../include/shm_addr.h"      /* shared memory pointer */
#include "../include/fscom.h"

void cls_snd();
void pname();
void rte_time();
static void logit0( char *msg, int ierr, char *who, char *type, char lsor);

logit_nds(msg,ierr,who,lsor)
char *msg;           /* a message to be logged, NULL if none */
int ierr;            /* error number, 0 if no error          */
char *who;           /* 2-char string identifying the error  */
char lsor;           /* char identifying source usually ':' or '/' */

{
  logit0(msg,ierr,who,"nd",lsor);
}
void logits(msg,ierr,who,lsor)
char *msg;           /* a message to be logged, NULL if none */
int ierr;            /* error number, 0 if no error          */
char *who;           /* 2-char string identifying the error  */
char lsor;           /* char identifying source usually ':' or '/' */

{
  logit0(msg,ierr,who,NULL,lsor);
}
void logit(msg,ierr,who)
char *msg;           /* a message to be logged, NULL if none */
int ierr;            /* error number, 0 if no error          */
char *who;           /* 2-char string identifying the error  */

{
  logit0(msg,ierr,who,NULL,'/');
}
void logit_nd(msg,ierr,who)
char *msg;           /* a message to be logged, NULL if none */
int ierr;            /* error number, 0 if no error          */
char *who;           /* 2-char string identifying the error  */

{
  logit0(msg,ierr,who,"nd",'/');
}
static void logit0(msg,ierr,who,type,lsor)
char *msg;           /* a message to be logged, NULL if none */
int ierr;            /* error number, 0 if no error          */
char *who;           /* 2-char string identifying the error  */
char *type;          /* data type NULL = "fs", "nd" = no display */
char lsor;           /* char identifying source usually ':' or '/' */

{
  char buf[1025];    /* Holds the complete log entry */
  char name[5];     /* The name of our main program */
  int it[6],ip1,ip2,l;
  char ssor[2];
 
/* First get the time and put dddhhmmss into the log entry.
*/
  rte_time(it,&it[5]);
  buf[0]='\0';
  int2str(buf,it[5],-4,1);
  strcat(buf,".");
  int2str(buf,it[4],-3,1);
  strcat(buf,".");
  int2str(buf,it[3],-2,1);
  strcat(buf,":");
  int2str(buf,it[2],-2,1);
  strcat(buf,":");
  int2str(buf,it[1],-2,1);
  strcat(buf,".");
  int2str(buf,it[0],-2,1);

/* For error messages, put ?ERROR xx nnnn into the log entry.
*/
  if (ierr != 0) {
    strcat(buf,"?ERROR ");
    strncat(buf,who,2);
    int2str(buf,ierr,-5,0);
  }
/* Get the name of our main program and append #pname# to the
   log entry. Then append the message.
*/
  else {
    pname(name);
    if(0!=memcmp(name,"     ",5)) {
      strcat(buf,"#");
      l=strlen(buf);
      memcpy(buf+l,name,5);
      buf[l+5]='\0';
      strcat(buf,"#");
    } else {
      ssor[0]=lsor;
      ssor[1]=0;
      strcat(buf,ssor);
    }
    if(msg!=NULL) {
      int n;
      int bufl=strlen(buf);
      int msgl=strlen(msg);
      n=sizeof(buf)-bufl-1;
      if(msgl < n)
	n=msgl;
      memcpy(buf+bufl,msg,n);
      buf[bufl+n]=0;
    } else
      strcat(buf,"empty message, program error");
  }
/* Send the complete log entry to ddout via class.
*/
  if(type == NULL)
    memcpy(&ip1,"fs",2);
  else
    memcpy(&ip1,type,2);
  memcpy(&ip2,"  ",2);
  if (ierr != 0) memcpy(&ip2,"b1",2);
/* for testing, send to output PLUS class */
/*  fprintf(stdout,"%s\n",buf); */
  cls_snd(&shm_addr->iclbox,buf,strlen(buf),ip1,ip2);

}

File: fslb/logit.f
*
* Copyright (c) 2020 NVI, Inc.
*
* This file is part of VLBI Field System
* (see http://github.com/nvi-inc/fs).
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
*
      subroutine logit(lmessg,nchar,lsor,lprocn,ierr,lwho,lwhat,nargs)

C  LOGIT formats a buffer for the log file and puts it into a
C  mailbox for DDOUT.
C  Character messages are displayed as is, errors are formatted
C  in a standard format. 
C  For normal messages, only 4 parameters need be input. 
C  For the command which terminates Field System operation,
C  the error parameter (5) is used as a cue to CLOUT to clear the class
C  after the message is processed. 
C  The error parameter (5) is also used by DDOUT as a signal 
C  to start a new log file.
C  If only 2 parameters are input, the calling program's name
C  prefaces the message. 
C  For error messages, all 6 (7 if LWHAT is present) parameters are
C  required although the message and procedure name are ignored. 
C 
      include '../include/fscom.i'
C 
C  INPUT: 
C 
      dimension lmessg(1) 
C      - the buffer holding the message 
C     NCHAR - length of the message, in characters
C     LSOR - source of the message, 1 character (determined by BOSS)
      dimension lprocn(1) 
C      - the procedure file from whence this message came, 12 chars 
C     IERR - error number, if this is an error. 
C            -999 indicates termination for CLOUT and DDOUT.
C     LWHO - source of the error, 2 chars, e.g. Qx for QUIKR or BO for BOSS.
C     LWHAT - what caused the error (usually a device name), 2 chars. 
C             If LWHAT is binary, it is converted to a 4-char ASCII number. 
C             Indicate that LWHAT is binary by setting LPROCN>0.
C             This parameter should be 0 or not present if not relevant.
C 
C  OUTPUT:  none
C 
C  LOCAL: 
C 
C     NCH - character counter in IBUF 
      integer*2 ibuf(MAX_CLS_MSG_I2)
C      - buffer in which log entry is formatted 
C     NARGS - number of arguments passed to us
      character*2 copt2
C     IOPT2/COPT2 - sent as optional parameter 2 on the class I/O.
C             "B1" indicates an error, a bell is used on the terminal display
C                  for negative values, positive values are warnings
C 
C  INITIALIZED: 
C 
C 
C 
C     1. Get the number of parameters we were sent. 
C     Set up the class I/O option.
C 
      iopt2 = 0 
      copt2 =' '
      if (nargs.gt.5) then
        copt2='b1'
      endif
      if (nargs.eq.5) then
        iopt2 = ierr
        ierr = 0
      end if
C                   If there are exactly 5 arguments, then
C                   transfer the IERR parameter directly. 
C 
C 
C     2. Call LOGEN to format the message.  Call it with the appropriate
C     number of parameters, depending on NARGS. 
C 
      if (nargs.eq.2) call logen4(ibuf,nch,lmessg,nchar) 
      if (nargs.eq.3) call logen5(ibuf,nch,lmessg,nchar,lsor)
      if (nargs.eq.4) call logen6(ibuf,nch,lmessg,nchar,lsor,lprocn) 
      if (nargs.eq.-4) call logen6d(ibuf,nch,lmessg,nchar,lsor,lprocn) 
      if (nargs.eq.5) call logen7(ibuf,nch,lmessg,nchar,lsor,lprocn, 
     .                ierr) 
      if (nargs.eq.6) call logen8(ibuf,nch,lmessg,nchar,lsor,lprocn, 
     .                ierr,lwho)
      if (nargs.eq.7) call logen9(ibuf,nch,lmessg,nchar,lsor,lprocn, 
     .                ierr,lwho,lwhat)
C 
C 
C     4. The buffer is formatted.  Pad with a blank for disk writing
C     purposes and put it into the mailbox.  That's all we do.
C 
      if(mod(nch,2).eq.1) then
         nch = ichmv_ch(ibuf,nch,' ')-2
C        Pad with a blank
      else
         nch=nch-1
      endif
c
      call fs_get_iclbox(iclbox)
      if(nargs.eq.5) then
        call put_bufi(iclbox,ibuf,-nch,'fs',iopt2) 
      else
        call put_buf(iclbox,ibuf,-nch,'fs',copt2) 
      endif
C 
      return
      end 
